File: validation/consumer.dart
import 'dart:convert';
import '../tmp/core/langs/dart/lib/ores_lib_core.dart' as core;
import '../tmp/baseline/langs/dart/lib/ores_lib_core.dart' as baseline;
import '../tmp/generated/corpus.dart';

void require(bool value, String label) {
  if (!value) throw StateError(label);
}
void rejects(void Function() action) {
  try { action(); } on FormatException { return; }
  throw StateError('invalid input was accepted');
}
// No dart:io and no assert statements; the VM and compiled JS execute identical checks.
void main() {
  final rows = jsonDecode(utf8.decode(base64Decode(encodedCases))) as List;
  require(rows.length > 3000, 'empty or truncated corpus');
  for (final row in rows) {
    final input = row['input'] as String;
    if (row['kind'] == 'correlation') {
      require(core.validCorrelationId(input) == row['expected'], row['id'] as String);
    } else if (row['expected'] == null) {
      rejects(() => core.normalizeEmailForRevocation(input));
    } else {
      final output = core.normalizeEmailForRevocation(input);
      require(output == row['expected'], row['id'] as String);
      require(core.normalizeEmailForRevocation(output) == output, 'normalization idempotence');
    }
  }
  for (final value in ['\ud800@example.com', 'a@\udfff.example', 'request-\ud800']) {
    rejects(() => core.normalizeEmailForRevocation(value));
    require(!core.validCorrelationId(value), 'unpaired surrogate');
  }
  final stored = List<int>.filled(32, 0);
  require(core.classifyIdempotency(null, stored) == core.IdempotencyDisposition.newRequest, 'new');
  for (var index = 0; index < 32; index += 1) {
    for (var byte = 0; byte < 256; byte += 1) {
      final incoming = List<int>.filled(32, 0)..[index] = byte;
      final expected = byte == 0 ? core.IdempotencyDisposition.replay : core.IdempotencyDisposition.conflict;
      require(core.classifyIdempotency(stored, incoming) == expected, 'digest result');
      require(stored.every((value) => value == 0), 'stored digest mutation');
      for (var j = 0; j < 32; j += 1) require(incoming[j] == (j == index ? byte : 0), 'incoming digest mutation');
    }
    for (final value in [-1, 256, 4294967296, -4294967296]) {
      final invalid = List<int>.filled(32, 0)..[index] = value;
      rejects(() => core.classifyIdempotency(null, invalid));
      rejects(() => core.classifyIdempotency(stored, invalid));
      rejects(() => core.classifyIdempotency(invalid, stored));
    }
  }
  for (final length in [0, 31, 33]) {
    final invalid = List<int>.filled(length, 0);
    rejects(() => core.classifyIdempotency(null, invalid));
    rejects(() => core.classifyIdempotency(invalid, stored));
  }
  // Positive controls for the old defects: no assertion that the old input was valid.
  require(baseline.normalizeEmailForRevocation('\u212a@example.com') == 'k@example.com', 'old Unicode defect must reproduce');
  require(baseline.classifyIdempotency(null, List<int>.filled(32, 256)) == baseline.IdempotencyDisposition.newRequest, 'old non-byte defect must reproduce');
  rejects(() => core.normalizeEmailForRevocation('\u212a@example.com'));
  rejects(() => core.classifyIdempotency(null, List<int>.filled(32, 256)));
  print(jsonEncode({'status': 'passed', 'sharedCorpusCases': rows.length, 'digestValuePositionCombinations': 8192, 'invalidDigestPlacements': 384, 'baselineDefectsReproduced': true}));
}

File: validation/consumer.rs
#![forbid(unsafe_code)]
use ores_lib_core::{classify_idempotency, normalize_email_for_revocation, valid_correlation_id, IdempotencyDisposition};

fn unhex(value: &str) -> String {
    assert_eq!(value.len() % 2, 0);
    let bytes: Vec<u8> = (0..value.len()).step_by(2)
        .map(|index| u8::from_str_radix(&value[index..index + 2], 16).expect("fixture byte"))
        .collect();
    String::from_utf8(bytes).expect("fixture Unicode scalar sequence")
}

#[test]
fn external_rust_api_replays_the_shared_corpus() {
    let mut count = 0;
    for line in include_str!("../tmp/generated/cases.tsv").lines() {
        let fields: Vec<_> = line.split('\t').collect();
        assert_eq!(fields.len(), 4);
        let input = unhex(fields[2]);
        match fields[1] {
            "email" if fields[3] == "-" => assert!(normalize_email_for_revocation(&input).is_err(), "{}", fields[0]),
            "email" => {
                let output = normalize_email_for_revocation(&input).expect("valid fixture rejected");
                assert_eq!(output, unhex(fields[3]), "{}", fields[0]);
                assert_eq!(normalize_email_for_revocation(&output).unwrap(), output);
            }
            "correlation" => {
                assert!(matches!(fields[3], "0" | "1"));
                assert_eq!(valid_correlation_id(&input), fields[3] == "1", "{}", fields[0]);
            }
            _ => panic!("unknown fixture kind"),
        }
        count += 1;
    }
    assert!(count > 3000);
    println!("shared corpus cases: {count}");
}

#[test]
fn every_digest_byte_at_every_position_matches_without_mutation() {
    let stored = [0_u8; 32];
    assert_eq!(classify_idempotency(None, &stored), IdempotencyDisposition::New);
    for index in 0..32 {
        for byte in 0..=255_u8 {
            let mut incoming = [0_u8; 32];
            incoming[index] = byte;
            let before = incoming;
            let expected = if byte == 0 { IdempotencyDisposition::Replay } else { IdempotencyDisposition::Conflict };
            assert_eq!(classify_idempotency(Some(&stored), &incoming), expected);
            assert_eq!(incoming, before);
            assert_eq!(stored, [0_u8; 32]);
        }
    }
}
